src/main/java/team/asd/service/ProductServiceImpl.java

```diff
@@ -0,0 +1,99 @@
+package team.asd.service;
+
+import lombok.NonNull;
+import org.apache.commons.collections4.CollectionUtils;
+import team.asd.constants.ProductState;
+import team.asd.entities.IsProduct;
+import team.asd.exceptions.WrongProductException;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+public class ProductServiceImpl implements IsProductService {
+	@Override
+	@NonNull
+	public List<String> defineProductNames(List<IsProduct> productList) throws WrongProductException {
+		if (CollectionUtils.isEmpty(productList)) {
+			return Collections.emptyList();
+		}
+		List<String> productNames = new ArrayList<>();
+		for (IsProduct p : productList) {
+			if (p == null) {
+				throw new WrongProductException("Product is null");
```

Comment:
You can add a separate validation method to check the product list for empty values and use this method in multiple places before actions with the collection. Or look for some null safe method in the CollectionUtils class.